src/dotnet/Fable.Compiler/Transforms/Replacements.fs

```diff
@@ -633,15 +628,18 @@ let rec equals (com: ICompiler) r equal (left: Expr) (right: Expr) =
         Helper.InstanceCall(left, "Equals", Boolean, [right]) |> is equal
     | Builtin(BclInt64|BclUInt64|BclBigInt as bt) ->
         Helper.CoreCall(coreModFor bt, "equals", Boolean, [left; right], ?loc=r) |> is equal
-    | ArrayOrList(modName, t) ->
+    | Array t ->
         let f = makeComparerFunction com t
-        Helper.CoreCall(modName, "equalsWith", Boolean, [f; left; right], ?loc=r) |> is equal
+        Helper.CoreCall("Array", "equalsWith", Boolean, [f; left; right], ?loc=r) |> is equal
+    | List _ ->
+        Helper.CoreCall("Util", "equals", Boolean, [left; right], ?loc=r) |> is equal
     | MetaType ->
         Helper.CoreCall("Reflection", "equals", Boolean, [left; right], ?loc=r) |> is equal
     | Tuple _ ->
         Helper.CoreCall("Util", "equalArrays", Boolean, [left; right], ?loc=r) |> is equal
-    | DeclaredType(ent,_) when hasBaseImplementingBasicMethods ent ->
-        Helper.InstanceCall(left, "Equals", Boolean, [right]) |> is equal
+    // unsafe optimization, left can sometimes be null
+    // | DeclaredType(ent,_) when hasBaseImplementingBasicMethods ent ->
+    //     Helper.InstanceCall(left, "Equals", Boolean, [right]) |> is equal
```

Comment:
Do you have an example when `left` may be null? `hasBaseImplementingBasicMethods` should only include records/unions/value types. And options are not DeclaredType in Fable AST?

Reply:
@alfonsogarciacaro Boxing a union type does it sometimes, with optimize-fcs on. I haven't looked into the details, since disabling this optimization has very little perf impact if any. We can come back to it in the future, but again, the perf difference is very small.